x-pack/plugins/upgrade_assistant/__jest__/client_integration/helpers/http_requests.ts

```diff
@@ -15,204 +15,132 @@ import {
   ResponseError,
 } from '../../../common/types';
 
+type HttpMethod = 'GET' | 'PUT' | 'DELETE' | 'POST';
+
 // Register helpers to mock HTTP Requests
-const registerHttpRequestMockHelpers = (server: SinonFakeServer) => {
-  const setLoadCloudBackupStatusResponse = (
-    response?: CloudBackupStatus,
-    error?: ResponseError
-  ) => {
-    const status = error ? error.statusCode || 400 : 200;
-    const body = error ? error : response;
+const registerHttpRequestMockHelpers = (
+  httpSetup: ReturnType<typeof httpServiceMock.createStartContract>,
+  shouldDelayResponse: () => boolean
+) => {
+  const mockResponses = new Map<HttpMethod, Map<string, Promise<unknown>>>(
+    ['GET', 'PUT', 'DELETE', 'POST'].map(
+      (method) => [method, new Map()] as [HttpMethod, Map<string, Promise<unknown>>]
+    )
+  );
+
+  const mockMethodImplementation = (method: HttpMethod, path: string) => {
+    const responsePromise = mockResponses.get(method)?.get(path) ?? Promise.resolve({});
+    if (shouldDelayResponse()) {
+      return new Promise((resolve) => {
+        setTimeout(() => resolve(responsePromise), 1000);
+      });
+    }
 
-    server.respondWith('GET', `${API_BASE_PATH}/cloud_backup_status`, [
-      status,
-      { 'Content-Type': 'application/json' },
-      JSON.stringify(body),
-    ]);
+    return responsePromise;
   };
 
-  const setLoadEsDeprecationsResponse = (response?: ESUpgradeStatus, error?: ResponseError) => {
-    const status = error ? error.statusCode || 400 : 200;
-    const body = error ? error : response;
+  httpSetup.get.mockImplementation((path) =>
+    mockMethodImplementation('GET', path as unknown as string)
+  );
+  httpSetup.delete.mockImplementation((path) =>
+    mockMethodImplementation('DELETE', path as unknown as string)
+  );
+  httpSetup.post.mockImplementation((path) =>
+    mockMethodImplementation('POST', path as unknown as string)
+  );
+  httpSetup.put.mockImplementation((path) =>
+    mockMethodImplementation('PUT', path as unknown as string)
+  );
 
-    server.respondWith('GET', `${API_BASE_PATH}/es_deprecations`, [
-      status,
-      { 'Content-Type': 'application/json' },
-      JSON.stringify(body),
-    ]);
+  const mockResponse = (method: HttpMethod, path: string, response?: unknown, error?: unknown) => {
```

Comment:
Love it! But was wondering if we should perhaps move all of this to a shared place (perhaps `kbn-test`) from which we could re-use this in other apps also? I think would also be cool if we could optionally specify headers too!

Reply:
I see your point, but the approach you use in UA and that was copied to few other plugins (additional abstraction over existing Core's mocks) is more like an exception and not something we (Platform) recommend. It's fine for a simple use case like you have, but if you need to customize response behavior (e.g. reply with different mock response based on parameters etc.) it becomes harder than using the Core's mock directly and also it's a bit harder to follow for people outside of the team.

So I'd rather localize this approach only inside plugins that opted to use it for now, until we have enough evidence that it can be useful for the majority of the plugins.

What do you think?

Reply:
Makes sense 👍🏼 , I think most of the platform-management apps use this pattern but let's keep an eye open for the other kibana apps and see if this is something more teams could benefit from.

Reply:
Yep, sounds good :+1: